contracts/cosmwasm-vm/cw-cluster-connection/src/helper.rs

```diff
@@ -0,0 +1,176 @@
+    }
+
+    pub fn hex_decode(&self, val: String) -> Result<Vec<u8>, ContractError> {
+        let hex_string_trimmed = val.trim_start_matches("0x");
+        hex::decode(hex_string_trimmed)
+            .map_err(|e| ContractError::InvalidHexData { msg: e.to_string() })
+    }
+
+    pub fn call_xcall_handle_message(
+        &self,
+        store: &dyn Storage,
+        nid: &NetId,
+        msg: Vec<u8>,
+    ) -> Result<SubMsg, ContractError> {
+        let xcall_host = self.get_xcall(store)?;
+        let xcall_msg = cw_xcall_lib::xcall_msg::ExecuteMsg::HandleMessage {
+            from_nid: nid.clone(),
+            msg,
+        };
+        let call_message: CosmosMsg<Empty> = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: xcall_host.to_string(),
+            msg: to_json_binary(&xcall_msg).unwrap(),
+            funds: vec![],
+        });
+        let sub_msg: SubMsg = SubMsg::reply_always(call_message, XCALL_HANDLE_MESSAGE_REPLY_ID);
+        Ok(sub_msg)
+    }
+
+    pub fn call_xcall_handle_error(
+        &self,
+        store: &dyn Storage,
+        sn: u128,
+    ) -> Result<SubMsg, ContractError> {
+        let xcall_host = self.get_xcall(store)?;
+        let xcall_msg = cw_xcall_lib::xcall_msg::ExecuteMsg::HandleError {
+            sn: sn.try_into().unwrap(),
+        };
+        let call_message: CosmosMsg<Empty> = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: xcall_host.to_string(),
+            msg: to_json_binary(&xcall_msg).unwrap(),
+            funds: vec![],
+        });
+        let sub_msg: SubMsg = SubMsg::reply_always(call_message, XCALL_HANDLE_ERROR_REPLY_ID);
+        Ok(sub_msg)
+    }
+
+    pub fn verify_signatures(
+        &self,
+        deps: Deps,
+        threshold: u8,
+        relayers: Vec<String>,
+        data: Vec<u8>,
+        signatures: Vec<Vec<u8>>,
+    ) -> Result<(), ContractError> {
+        if signatures.len() < threshold.into() {
+            return Err(ContractError::InsufficientSignatures);
+        }
+
+        let message_hash = sha256(&data);
+
+        let mut signers: HashMap<String, bool> = HashMap::new();
+
+        for signature in signatures {
+            if signature.len() != 65 {
+                return Err(ContractError::InvalidSignature);
+            }
+            match deps
+                .api
+                .secp256k1_recover_pubkey(&message_hash, &signature[0..64], signature[64])
+            {
+                Ok(pubkey) => {
+                    let pk = VerifyingKey::from_sec1_bytes(&pubkey)
+                        .map_err(|_| ContractError::InvalidSignature)?;
+
+                    let pk_hex = hex::encode(pk.to_bytes());
+                    if relayers.contains(&pk_hex) && !signers.contains_key(&pk_hex) {
+                        signers.insert(pk_hex, true);
+                        if signers.len() >= threshold.into() {
+                            return Ok(());
+                        }
+                        break;
```

Comment:
why are we doing break here ? wont it stop loop in first iteration?

Reply:
my bad.